### PR TITLE
Make dot more generic (and safer)

### PR DESCRIPTION
**Pull request by joehuchette**
Also adds a missing method so that e.g. ``+(::OneIndexedArray, ::OneIndexedArray)`` works as expected.

### CODE REVIEW

**Comment by joehuchette** [comment]
SCS :'(

File: src/operators.jl [hunk 1]
```diff
@@ -241,26 +243,25 @@ function Base.sum{S,T}(affs::Array{GenericAffExpr{S,T}})
...
-    size(lhs.innerArray) == size(rhs.innerArray) || error("Incompatible dimensions")
-    return QuadExpr(vec(lhs.innerArray), vec(rhs.innerArray), ones(length(lhs.innerArray)), AffExpr())
+Base.dot{T,S,N}(lhs::OneIndexedArray{T,N},rhs::OneIndexedArray{S,N}) = _dot(lhs.innerArray, rhs.innerArray)
+Base.dot{T,S,N}(lhs::OneIndexedArray{T,N},rhs::Array{S,N}) = _dot(lhs.innerArray, rhs)
+Base.dot{T,S,N}(lhs::Array{T,N},rhs::OneIndexedArray{S,N}) = _dot(lhs, rhs.innerArray)
+Base.dot{T<:JuMPTypes,S,N}(lhs::Array{T,N},rhs::Array{S,N}) = _dot(lhs,rhs)
+Base.dot{T<:JuMPTypes,S<:JuMPTypes,N}(lhs::Array{T,N},rhs::Array{S,N}) = _dot(lhs,rhs)
+Base.dot{T,S<:JuMPTypes,N}(lhs::Array{T,N},rhs::Array{S,N}) = _dot(lhs,rhs)
+
+function _dot{T,S}(lhs::Array{T}, rhs::Array{S})
```
Comment (mlubin) [inline src/operators.jl]: This is slow in every case, can't we do better? At least ``dot(Vector{Float64},OneIndexedArray{Variable})`` should be very fast.

Comment (joehuchette) [inline src/operators.jl]: I'd rather just get rid of multidimensional dot, to be honest, considering
it's not in Base.

Comment (tkelman) [inline src/operators.jl]: there is now a `vecdot` in 0.4

Comment (joehuchette) [inline src/operators.jl]: Cool, I thought I remembered something about to be added. I'll change the naming then.

**Pull request merged**